File: Parser/Command.cs
using System;
using System.ComponentModel.Design;

public enum Opcode
{
	Create,
	Agent
}
public enum Instruction
{
	Play,
	Go,
	Stop,
	Agents
}
public enum Action
{
	Clap,
	Dance
}
public class Command
{
	private Opcode opcode;
	private int quantity;
	private Instruction instruction;
	private Action action;
	private String details;
	private String toString;
	public Command(String command)
	{
		toString = "";
		// Create x agents {(a,b,c), (m,n,p), ... (d,e,f)}
		if(command.Substring(0, 6).Equals("Create", StringComparison.InvariantCultureIgnoreCase))
        {
			opcode = Opcode.Create;
			toString += "Create ";

			String number = "";
			int i;
			for (i = 7; command[i] == " "; i++) { number += command[i];}
			quantity = int.Parse(number);
			toString += quantity + " ";

			if (command.Substring((i + 1), 6).Equals("agents", StringComparison.InvariantCultureIgnoreCase))
            {
				instruction = Instruction.Agents;
				i += 7;
			}

			details = command.Substring(i);

			toString += details;
		}
		// Agent x go to (a,b, c)
		else if (command.Substring(0,5).Equals("Agent", StringComparison.InvariantCultureIgnoreCase))
        {
			opcode = Opcode.Agent;
			toString += "Agent ";

			String number = "";
			int i;
			for (i = 6; command[i] == " "; i++) { number += command[i]; }
			quantity = int.Parse(number);
			toString += quantity + " ";

			if(command.Substring(i + 1, 5).Equals("go to", StringComparison.InvariantCultureIgnoreCase))
            {
				instruction = Instruction.Go;
				i += 6;
				toString += "Go ";
            }
			if(command.Substring(i + 1, 4).Equals("stop", StringComparison.InvariantCultureIgnoreCase))
            {
				instruction = Instruction.Go;
				i += 5;
				toString += "Stop ";
            }
			if (command.Substring(i + 1, 4).Equals("paly", StringComparison.InvariantCultureIgnoreCase))
			{
				instruction = Instruction.Play;
				i += 5;
				toString += "Play ";

				if (command.Substring(i, 5).Equals("claps", StringComparison.InvariantCultureIgnoreCase)){
					action = Action.Clap;
					i += 6;
					toString += "Clap ";
				}
				if (command.Substring(i, 5).Equals("dance", StringComparison.InvariantCultureIgnoreCase)){
					action = Action.Dance;
					i += 6;
					toString += "Dance ";
				}
			}

			details = command.Substring(i);
			toString += details;
		}
	}

	public Opcode GetOpcode()
    {
		return opcode;
    }

	// to sring
	public override string ToString()
    {
		return toString;
    }
}

File: Parser/Command.cpp
#include <iostream>
#include <string>
#include <algorithm>

using namespace std;


class Command{

	Command(string command){

		toString = "";

		//Start
		if(iequals(command.substr (0,5), "Start")){

			opcode = Opcode::START;

			//in
			if(iequals(command.substr (6,2), "in"){

				instruction = Instruction::IN;
				place = command.substr (9);

			}
		}
		//Begin
		else if(iequals(command.substr (0,5), "Begin")){

			opcode = Opcode::BEGIN;

			//Loop
			if(iequals(command.substr (5,4), "Loop")){

				instruction = Instruction::LOOP;
			}
		}
		//Go
		else if(iequals(command.substr (0,2), "Go")){

			opcode = Opcode::GO;

			//To
			if(iequals(command.substr (2,2), "To")){

				instruction = Instruction::TO;
				place = command.substr(6);
			}
		}
		//Play
		else if(iequals(command.substr (0,4), "Play")){
			
			opcode = Opcode::PLAY;

			//Claps
			if(iequals(command.substr (5,5), "Claps")){
				
				instruction = Instruction::CLAPS;

			}

			//Dance
			if(iequals(command.substr (5,5), "Dance")){
				
				instruction = Instruction::DANCE;

			}
		}
		//End
		else if(iequals(command.substr (0,3), "End")){

			opcode = Opcode::END;

			//Loop
			if(iequals(command.substr (3,4), "Loop")){

				instruction = Instruction::LOOP;
			}
		}
	}

	bool iequals(const string& a, const string& b)
	{
		return std::equal(a.begin(), a.end(), b.begin(),
						[](char a, char b) {
							return tolower(a) == tolower(b);
						});
	}
}
